File: ffi/ts/src/keys/privSpendingKey.ts
import {
  calcPrivSpendingKey,
} from '../blsct'

import { Scalar } from '../scalar'
import { PublicKey } from './publicKey'

/** Represents a private spending key. A private spending key is a `Scalar` and introduces no new functionality; it serves purely as a semantic alias.
  *
  * Examples:
  * ```ts
  * const { PrivSpendingKey, PublicKey, SpendingKey } = require('navio-blsct')
  * const pk = PublicKey.random()
  * const vk = Scalar.random()
  * const sk = SpendingKey.random()
  * new PrivSpendingKey(pk, vk, sk, 1, 2)
  * ```
  */
export class PrivSpendingKey extends Scalar {
  /** Constructs a new `PrivSpendingKey` instance.
   *
   * @param blindingPubKey - The public key used for blinding.
   * @param viewKey - The view key.
   * @param spendingKey - The spending key.
   * @param account - The account.
   * @param address - The address.
   */
  constructor(
    blindingPubKey: PublicKey,
    viewKey: Scalar,
    spendingKey: Scalar,
    account: number,
    address: number
  ) {
    const blsctPsk = calcPrivSpendingKey(
      blindingPubKey.value(),
      viewKey.value(),
      spendingKey.value(),
      account,
      address
    )
    super(blsctPsk)
  }
}

File: ffi/rust/src/keys/priv_spending_key.rs
use crate::{
  ffi::calc_priv_spending_key,
  keys::{
    child_key_desc::tx_key_desc::spending_key::SpendingKey,
    public_key::PublicKey,
  },
};

crate::macros::impl_key!(PrivSpendingKey);

impl PrivSpendingKey {
  pub fn new(
    blinding_pub_key: PublicKey,
    view_key: Scalar,
    spending_key: SpendingKey,
    account: i64,
    address: u64,
  ) -> Self {
    let priv_spending_key = unsafe {
      calc_priv_spending_key(
        blinding_pub_key.value(),
        view_key.value(),
        spending_key.value(),
        account,
        address,
      )
    };
    let obj = BlsctObj::from_c_obj(priv_spending_key);
    obj.into()
  }
}

#[cfg(test)]
mod tests {
  use super::*;
  use crate::{
    initializer::init,
    keys::child_key::ChildKey,
  };

  #[test]
  fn test_new() {
    init();
    let child_key = ChildKey::random().unwrap();
    let tx_key = child_key.to_tx_key();
    let view_key = tx_key.to_view_key();
    let spending_key = tx_key.to_spending_key();
    let blinding_pub_key = PublicKey::random().unwrap();

    let _: PrivSpendingKey = PrivSpendingKey::new(
      blinding_pub_key,
      view_key,
      spending_key,
      123,
      456,
    );
  }
}
